rfd/0095-teleport-key-agent-extension.md

```diff
@@ -0,0 +1,251 @@
+        byte[]          digest
+        string          hash name
+        string          salt length
+
+The resulting signature will be returned alongside the signing algorithm used.
+
+        byte            SSH_AGENT_SIGN_RESPONSE
+        string          signature format
+        byte[]          signature blob
+
+Hash name should be the stringified representation of a golang `crypto.Hash` value. For example, `SHA-1`, `SHA-256`, or `SHA-512`.
+
+Salt length can be empty for algorithms that don't use a salt, or a positive integer for those that do (such as `RSAPSS`). Salt length can also be set to `auto` to automatically use the largest salt length possible during signing, which can be auto-detected during verification.
+
+**Prompt MFA Challenge Extension:**
+
+This extension can be used to issue an MFA challenge prompt to the user's local machine, which enables support for MFA functionality including `tsh` MFA login and per-session MFA verification.
+
+        byte            SSH_AGENTC_EXTENSION
+        string          prompt-mfa-challenge@goteleport.com
+        string          proxy address
+        []byte          challenge blob (json)
+
+Where challenge blob is a json encoded `api/client/proto.MFAAuthenticateChallenge`.
+
+The resulting challenge response can then be used for MFA verification.
+
+        byte            SSH_AGENT_SUCCESS
+        []byte          challenge response blob (json)
+
+Where challenge response blob is a json encoded `api/client/proto.MFAAuthenticateResponse`.
+
+Note: The protobuf structs used above are subject to change, which may lead to backwards compatibility concerns. In this case, this extension should be taken into consideration before making changes to these structs.
+
+### Additional Considerations
+
+#### Syncing local/remote `tsh` profiles
+
+Instead of only forwarding a single `tsh` profile in a remote forwarding session, we could forward a user's entire `~/.tsh` and keep their local and remote `tsh` profile synced. This means that the remote session would follow the local `tsh` profile through profile switches, log outs, new log ins, etc. It could even be possible to perform these actions from the remote session. This experience would also be more in line with OpenSSH agent forwarding, which forwards all keys in the user's `~/.ssh` directory.
+
+While this would provide the most seamless UX experience between local and remote sessions, this approach would increase the security risk of local agent forwarding. For example, if a user two active `tsh` profiles, one for administration and one for basic operations, the user could `tsh ssh -o "ForwardAgent local"` into a widely accessible server. The user's administrative `tsh` profile could potentially be used through their forward agent connection, due to the inherent risks of ssh agent forwarding.
+
+If we were to enable this functionality, we would need to do one or both of the following:
```

Comment:
Also, users should be informed about the increased security risk if they would forward an entire ~/.tsh